Move unit tests to separate directories, and document (#8886)
NixOS/nix: doc/manual/src/contributing/testing.md

## Unit-tests

The unit-tests for defined using the [googletest] and [rapidcheck] frameworks.

[googletest]: https://google.github.io/googletest/
[rapidcheck]: https://github.com/emil-e/rapidcheck

### Source and header layout

> An example for `libnixexpr`, demonstrating much of what is described below
>
> ```
> src
> ├── libexpr
> │   ├── value/context.hh
> │   ├── value/context.cc
> │   │
> │   …
> ├── libexpr-tests
> │   ├── tests/
> │   │   ├── value/context.hh
> │   │   │
> │   │   …
> │   ├── value/context.cc
> │   │
> │   …
> …
> ```

The tests for each Nix library (`libnixexpr`, `libnixstore`, etc..) live inside a directory `src/${library_name_without-nix}-tests` next to the directory for the library (`src/${library_name_without-nix}`).

> Given a `foo/bar.{hh,cc}` interface (header) and implementation pair in the original library, we write tests for it in `foo/bar.cc`, and (possibly) declare additional interfaces in `tests/foo/bar.cc` in the test directory.

Reviewer: Rather than the arbitrary `foo` and `bar` here, could you maybe reword this so it picks up the naming in the tree above? That would make it much more concrete, and much easier for me to understand.

Specifically, why is there a `libexpr-tests/value/context.cc` (I imagine this is just the tests for `libexpr/value/context.cc`) as well as a `libexpr-tests/tests/value/context.hh` (what's the difference between this and `libexpr/value/context.hh`)?

Reviewer 2: the fix here is to prevent name collisions in the first place by separating out the directories, its less a case of concerning the contents of the files.

Author: ```suggestion
Given a interface (header) and implementation pair in the original library, say, `value/context.{hh,cc}`, we write tests for it in `value/context.cc`, and (possibly) declare additional interfaces for testing purposes in `test/value/context.hh`.
(Note only the extra header is nested inside the `test` directory. This is explained further below.) 
```
How's this?

Reviewer: Yes, that is much easier for me to wrap my head around! Thanks!

## Unit-tests

> The unit-tests for defined using the [googletest] and [rapidcheck] frameworks.

Reviewer 3: ```suggestion
The unit-tests are defined using the [googletest] and [rapidcheck] frameworks.
```

[googletest]: https://google.github.io/googletest/
[rapidcheck]: https://github.com/emil-e/rapidcheck

### Source and header layout

> An example for `libnixexpr`, demonstrating much of what is described below
>
> ```
> src
> ├── libexpr
> │   ├── value/context.hh
> │   ├── value/context.cc
> │   │
> │   …
> ├── libexpr-tests
> │   ├── tests/
> │   │   ├── value/context.hh
> │   │   │
> │   │   …
> │   ├── value/context.cc
> │   │
> │   …
> …
> ```

> The tests for each Nix library (`libnixexpr`, `libnixstore`, etc..) live inside a directory `src/${library_name_without-nix}-tests` next to the directory for the library (`src/${library_name_without-nix}`).

Reviewer 3: Scrolling back and forth I understand what you mean by "without-nix", but would it hurt so much to not have that distinction? E.g. by removing the nix in the original library name or adding it for the tests as well? It makes something that's fairly involved and obscure needlessly hard to grok.

Reviewer 2: that was actually discussed as a future goal of current refactoring efforts :) the problem currently is striking a balance between discrete specific small improvements that can be introduced overtime, and a just one huge commit that changes everything. seeing as the current layout uses libnix* to discuss nix libraries, it would be best to keep as is for now imo

Author: Yeah the other dir is `libexpr` so it seems prudent to let that be for now and call this `libexpr-tests`

> Given a interface (header) and implementation pair in the original library, say, `value/context.{hh,cc}`, we write tests for it in `value/context.cc`, and (possibly) declare additional interfaces for testing purposes in `test/value/context.hh`.

Reviewer 3: ```suggestion
Given a interface (header) and implementation pair in the original library, say, `libexpr/value/context.{hh,cc}`, we write tests for it in `libexpr-tests/value/context.cc`, and (possibly) declare additional interfaces for testing purposes in `libexpr-tests/tests/value/context.hh`.
```

Is it test or tests now? I'd prefer test because it's fewer characters and doesn't lose relevant information.

Author: Oh oops, should be `test`

> (Note only the extra header is nested inside the `test` directory. This is explained further below.)

Reviewer 3: ```suggestion

> **Note**
>
> Only the extra header is nested inside the `test` directory. This is explained further below.
```

If you have an anchor there you can link directly.

> The use of a separate directory for the unit tests might seem inconvenient, as the tests are not "right next to" the part of the code they are testing.

Reviewer 3: ```suggestion
### Rationale

The use of a separate directory for the unit tests might seem inconvenient, as the tests are not "right next to" the part of the code they are testing.
```

Breaking this long text down with headings will make it easier to navigate. You can use the leading questions as level 4 headers.

But organizing the tests this way has one big benefit:
there is no risk of any build-system wildcards for the library accidentally picking up test code that shoud not built and installed as part of the library.

Likewise, the use of the `test/` subdir might seem superfluous:
Isn't the point of the `*-test` subdir to indicate that these files are tests?

> Why do we need another `test` subdir too?

Reviewer 3: ```suggestion
Why do we need another `test` subdirectory?
```

The answer is that we need to be able to tell apart the two headers, and make sure we include the right one.
For `.cc` files, this is matter of doing `#include "foo.hh"` vs `#include "tests/foo.hh`
For `.hh` files, this is a bit more subtle, because a `#include "foo.hh` instead `tests/foo.hh` will end up including itself because `#include "..."`
[always prioritizes files in the same directory](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/c99.html#tag_20_11_04), rather than the original header from the library.
Instead, use `#include <foo.hh>` to get the original library header, and `#include "tests/foo.hh"` to get the test header.

Why do we have test headers at all?
These are usually for [rapidcheck]'s `Arbitrary` machinery, which is used to describe how to generate values of a given type for the sake of running property tests.

> Because types contain other types, `Arbitrary "instances" for some type are not just useful for testing that type, but also any other type that contains it.

Reviewer 3: ```suggestion
Because types contain other types, `Arbitrary` "instances" for some type are not just useful for testing that type, but also any other type that contains it.
```

> (Indeed, if we don't reuse the upstream type's `Arbitrary` instance, the downstream type's `Arbitrary` instance would become much more complex and hard to understand.)

Reviewer 3: ```suggestion
Indeed, if we don't reuse the upstream type's `Arbitrary` instance, the downstream type's `Arbitrary` instance would become much more complex and hard to understand.
```

This is a proper argument, no need to parenthesise.

> ├── libexpr
> │   ├── value/context.hh
> │   ├── value/context.cc
> │   …
> │
> ├── tests
> │   │
> │   …
> │   └── unit
> │       ├── libutil
> │       │   │
> │       │   …
> │       │
> │       ├── libexpr
> │       │   └── test
> │       │       ├── value/context.hh
> │       │       ├── value/context.cc
> │       │       …
> │       …
> │       └── data
> │           ├── libutil
> │           │   ├── git/tree.txt
> │           │   …
> │           ├── libexpr
> │           │   │
> │           │   …
> │           …
> …
> ```

> The tests for each Nix library (`libnixexpr`, `libnixstore`, etc..) live inside a directory `src/${library_name_without-nix}-tests` next to the directory for the library (`src/${library_name_without-nix}`).

Reviewer 4: This is now wrong, right?

Author: Good catch. There are a few things that need to be fixed here.

The answer is that we need to be able to tell apart the two headers, and make sure we include the right one.
For `.cc` files, this is matter of doing `#include "foo.hh"` vs `#include "tests/foo.hh`
For `.hh` files, this is a bit more subtle, because a `#include "foo.hh` instead `test/foo.hh` will end up including itself because `#include "..."`
[always prioritizes files in the same directory](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/c99.html#tag_20_11_04), rather than the original header from the library.
Instead, use `#include <foo.hh>` to get the original library header, and `#include "tests/foo.hh"` to get the test header.

Why do we have test headers at all?
These are usually for [rapidcheck]'s `Arbitrary` machinery, which is used to describe how to generate values of a given type for the sake of running property tests.
Because types contain other types, `Arbitrary` "instances" for some type are not just useful for testing that type, but also any other type that contains it.
Indeed, if we don't reuse the upstream type's `Arbitrary` instance, the downstream type's `Arbitrary` instance would become much more complex and hard to understand.

> In order to reuse these instances, we therefore declare them in these testing headers.

Reviewer 4: Is don't think we care a bout that at all here. This is nice for a commit message, but users/contributors generally shouldn't be really concerned about it, and wouldn't read it any way.

Reviewer 3: I think it's the generally he right place for this kind of contextual information, but we may consider moving that out of the manual and into the tree.

Author: I want contributors to also write property tests and "Arbitrary instances", so I do want that pattern to be documented.

Author: Amid many other changes this text is now reorganized out of the file layout section. Hope that is better.
